### DISCUSSION TITLE
[New Feature] Support array:some and array:every langlib functions

### DISCUSSION THREAD

**Discussion by KavinduZoysa** [ballerina-platform/ballerina-lang#35481] [General]
```
# Tests whether a function returns true for some member of an array.
# func` is called for each member of `arr` in order unless and until a call returns true.
# + arr - the array
# + func - function to apply to each member
# + return - true if func returns true for some member of `arr`; otherwise, false
# 
public isolated function some(Type[] arr, @isolatedParam function(Type val) returns boolean func) returns boolean = external;

# Tests whether a function returns true for every member of an array.
# `func` is called for each member of `arr` in order unless and until a call returns false.
# + arr - the array
# + func - function to apply to each member
# + return - true if func returns true for every member of `arr`; otherwise, false
# 
public isolated function every(Type[] arr, @isolatedParam function(Type val) returns boolean func) returns boolean = external;
```

These are the two functions we need to support (Please check spec issue https://github.com/ballerina-platform/ballerina-spec/issues/1005). For both cases, we can implement ballerina functions in array.bal.

**array:some**

This langlib function will take an array and a function pointer as parameters and returns a boolean. 
In the implementation, we have to iterate over the elements in the input array and call the input function for each element. Therefore the input of the function should be an array element and it should return a boolean. If that function returns true we have to stop the iteration and return true. Please find the below flow diagram for a clear understanding.

<img src="https://user-images.githubusercontent.com/28332708/158527285-f1a07a01-d31b-483e-8cab-3ce6e928bbd4.png"  width="500" height="700" />

**array:every**

The signature of this function is the same as the above function. 
If the input function returns false we have to stop the iteration and return false. Please find the below flow diagram for a clear understanding.

<img src="https://user-images.githubusercontent.com/28332708/158527365-e0b5e5e5-613f-4721-adb3-4afb015dfbc7.png"  width="500" height="700" />

Highly appreciate your input on this discussion.

**Comment by MaryamZi** [comment]
> These are the two functions we need to support (Please check spec issue [ballerina-platform/ballerina-spec#1005](https://github.com/ballerina-platform/ballerina-spec/issues/1005)). For both cases, we can implement ballerina functions in array.bal.
> 

Minor detail - we "**need to** implement" this in array.bal (the array langlib), rather than "**can** implement". In the array langlib so that it is an array langlib function, in array.bal so that we are maintaining a 1:1 mapping with the spec.

> <img alt="" width="500" height="700" src="https://user-images.githubusercontent.com/28332708/158527285-f1a07a01-d31b-483e-8cab-3ce6e928bbd4.png">
>

`index++` needs to happen before the the comparison with array length? And after incrementing needs to do `call func(arr[index])`? 

> <img alt="" width="500" height="700" src="https://user-images.githubusercontent.com/28332708/158527365-e0b5e5e5-613f-4721-adb3-4afb015dfbc7.png">
> 

Same comment as for `array:some`.

**Reply by KavinduZoysa** [reply]
Yes, that is correct

**Comment by KavinduZoysa** [comment]
Please find the progress of the `array:some` implementation in PR https://github.com/ballerina-platform/ballerina-lang/pull/35496

**Comment by MaryamZi** [comment]
Also see https://github.com/ballerina-platform/ballerina-lang/pull/35531#discussion_r832840598 regarding `array:every` returning false when the list is empty.
